File: python/problem4.py
"""
Link:
    https://projecteuler.net/problem=4

Problem:
    A palindromic number reads the same both ways. The largest palindrome made from the
    product of two 2-digit numbers is 9009 = 91 × 99.

    Find the largest palindrome made from the product of two 3-digit numbers.
"""

from math import log10

# function to reverse a number with math
def reverse_num(x):
    decs = [10**n for n in range(0, int(log10(x)+1))]
    nums = [m % 10 for m in [x // n for n in decs]]
    reversed_num = sum([n * m for n, m in zip(nums, decs[::-1])])
    return reversed_num

if __name__ == "__main__":
    res = 0

    for n in range(999, 100, -1):
        for m in range(999, 100, -1):
            z = n * m

            if res < z == reverse_num(z):
                res = z

    # result
    print(res)

File: c/problem4.c
#include <stdio.h>
#include <math.h>

unsigned int ReverseNumber(double x);

/*	https://projecteuler.net/problem=4	*/
int main()
{
	unsigned int i, j;
	unsigned int res = 0;

	for( i = 999; i >= 100; i-- )
	{
		for ( j = 999; j >= 100; j-- )
		{
			unsigned int prod = i * j;

			if (prod > res && ReverseNumber(prod) == prod )
			{
				res = prod;
			}
		}
	}

	// result
	printf("%d", res);

	return 0;
}

/* function to reverse a number with math */
unsigned int ReverseNumber(double x)
{
	unsigned int i, j;
	unsigned int len = (unsigned int) log10(x);
	unsigned int decs[len];

	j = 0;
	for ( i = len; i > 0; i-- )
	{
		decs[ j++ ] = pow( 10, i );
	}
	decs[ len ] = pow( 10, 0 );

	unsigned int sum = 0;
	for ( int i = 0; i <= len; i++ )
	{
		unsigned int y = x / decs[ j-- ];
		y %= 10;
		y *= decs[ i ];
		sum = sum + y;
	}

	return sum;
}


